src/storage/src/hummock/iterator/forward_user.rs

```diff
@@ -96,62 +132,34 @@ impl<I: HummockIterator<Direction = Forward>> UserIterator<I> {
     }
 
     /// Gets the iterator move to the next step.
+    /// See `IteratorState` for the state machine details.
     ///
     /// Returned result:
     /// - if `Ok(())` is returned, it means that the iterator successfully move to the next position
     ///   (may reach to the end and thus not valid)
     /// - if `Err(_) ` is returned, it means that some error happened.
     pub async fn next(&mut self) -> HummockResult<()> {
-        while self.iterator.is_valid() {
-            let full_key = self.iterator.key();
-            let epoch = full_key.epoch_with_gap.pure_epoch();
-
-            // handle multi-version
-            if epoch < self.min_epoch || epoch > self.read_epoch {
-                self.iterator.next().await?;
-                continue;
-            }
+        // Move the iterator to the next step if it is currently potined to a ready entry.
+        if self.state.is_ready() {
+            self.state.transition_to(IteratorState::Next);
+        }
 
-            if self.last_key.user_key.as_ref() != full_key.user_key {
-                // It is better to early return here if the user key is already
-                // out of range to avoid unnecessary access on the range tomestones
-                // via `delete_range_iter`.
-                // For example, if we are iterating with key range [0x0a, 0x0c) and the
-                // current key is 0xff, we will access range tombstones in [0x0c, 0xff],
-                // which is a waste of work.
-                if self.key_out_of_range() {
-                    self.out_of_range = true;
-                    return Ok(());
+        // Loop until the iterator becomes ready or invalid.
+        loop {
+            match self.state {
+                IteratorState::Init => {
```

Comment:
The state `Init` and `Next` only affects the logic here. If we do the following change, the two states can be merged to a state named like `Intermediate`.
```
async fn rewind() {
    self.last_key = FullKey::default();
    self.advance_to_valid().await;
}

async fn next() {
    self.iterator.next().await;
    self.advance_to_valid().await;
}

async fn advance_to_valid() {
     while self.state is not intermediate {
          self.step_one().await;
      }
}


```

Reply:
Good idea!

Reply:
Fixed.